// amis/lib/index.js
define('node_modules/amis-editor/node_modules/amis/lib/index', function(require, exports, module) {

  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.makeTranslator = exports.registerLocale = exports.setDefaultLocale = exports.getDefaultLocale = exports.classnames = exports.getClassPrefix = exports.classPrefix = exports.getTheme = exports.theme = exports.setDefaultTheme = exports.validateObject = exports.validate = exports.ScopedContext = exports.Scoped = exports.registerIcon = exports.Icon = exports.setIconVendor = exports.resolveVariableAndFilter = exports.resolveVariable = exports.relativeValueRe = exports.filterDate = exports.filterSchema = exports.resolveRenderer = exports.getRendererByName = exports.normalizeOptions = exports.str2rules = exports.addRule = exports.setCustomEvalExpression = exports.setCustomEvalJs = exports.evalJS = exports.evalExpression = exports.reigsterTplEnginer = exports.registerFilter = exports.resizeSensor = exports.utils = exports.filter = exports.buildApi = exports.wrapFetcher = exports.OptionsControl = exports.FormItem = exports.RegisterStore = exports.Renderer = exports.updateEnv = exports.clearStoresCache = exports.render = void 0;
  var tslib_1 = require("node_modules/tslib/tslib");
  /** @license amis v1.0.14
   *
   * Copyright Baidu
   *
   * This source code is licensed under the Apache license found in the
   * LICENSE file in the root directory of this source tree.
   */
  var factory_1 = require("node_modules/amis-editor/node_modules/amis/lib/factory");
  Object.defineProperty(exports, "render", { enumerable: true, get: function () { return factory_1.render; } });
  Object.defineProperty(exports, "Renderer", { enumerable: true, get: function () { return factory_1.Renderer; } });
  Object.defineProperty(exports, "getRendererByName", { enumerable: true, get: function () { return factory_1.getRendererByName; } });
  Object.defineProperty(exports, "resolveRenderer", { enumerable: true, get: function () { return factory_1.resolveRenderer; } });
  Object.defineProperty(exports, "filterSchema", { enumerable: true, get: function () { return factory_1.filterSchema; } });
  Object.defineProperty(exports, "clearStoresCache", { enumerable: true, get: function () { return factory_1.clearStoresCache; } });
  Object.defineProperty(exports, "updateEnv", { enumerable: true, get: function () { return factory_1.updateEnv; } });
  var api_1 = require("node_modules/amis-editor/node_modules/amis/lib/utils/api");
  Object.defineProperty(exports, "wrapFetcher", { enumerable: true, get: function () { return api_1.wrapFetcher; } });
  Object.defineProperty(exports, "buildApi", { enumerable: true, get: function () { return api_1.buildApi; } });
  var tpl_1 = require("node_modules/amis-editor/node_modules/amis/lib/utils/tpl");
  Object.defineProperty(exports, "filter", { enumerable: true, get: function () { return tpl_1.filter; } });
  Object.defineProperty(exports, "reigsterTplEnginer", { enumerable: true, get: function () { return tpl_1.reigsterTplEnginer; } });
  Object.defineProperty(exports, "evalExpression", { enumerable: true, get: function () { return tpl_1.evalExpression; } });
  Object.defineProperty(exports, "evalJS", { enumerable: true, get: function () { return tpl_1.evalJS; } });
  Object.defineProperty(exports, "setCustomEvalJs", { enumerable: true, get: function () { return tpl_1.setCustomEvalJs; } });
  Object.defineProperty(exports, "setCustomEvalExpression", { enumerable: true, get: function () { return tpl_1.setCustomEvalExpression; } });
  var utils = tslib_1.__importStar(require("node_modules/amis-editor/node_modules/amis/lib/utils/helper"));
  exports.utils = utils;
  var resize_sensor_1 = require("node_modules/amis-editor/node_modules/amis/lib/utils/resize-sensor");
  Object.defineProperty(exports, "resizeSensor", { enumerable: true, get: function () { return resize_sensor_1.resizeSensor; } });
  var IconPickerIcons_1 = require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/IconPickerIcons");
  Object.defineProperty(exports, "setIconVendor", { enumerable: true, get: function () { return IconPickerIcons_1.setIconVendor; } });
  var icons_1 = require("node_modules/amis-editor/node_modules/amis/lib/components/icons");
  Object.defineProperty(exports, "Icon", { enumerable: true, get: function () { return icons_1.Icon; } });
  Object.defineProperty(exports, "registerIcon", { enumerable: true, get: function () { return icons_1.registerIcon; } });
  var store_1 = require("node_modules/amis-editor/node_modules/amis/lib/store/index");
  Object.defineProperty(exports, "RegisterStore", { enumerable: true, get: function () { return store_1.RegisterStore; } });
  var locale_1 = require("node_modules/amis-editor/node_modules/amis/lib/locale");
  Object.defineProperty(exports, "setDefaultLocale", { enumerable: true, get: function () { return locale_1.setDefaultLocale; } });
  Object.defineProperty(exports, "getDefaultLocale", { enumerable: true, get: function () { return locale_1.getDefaultLocale; } });
  Object.defineProperty(exports, "makeTranslator", { enumerable: true, get: function () { return locale_1.makeTranslator; } });
  Object.defineProperty(exports, "registerLocale", { enumerable: true, get: function () { return locale_1.register; } });
  // 注册渲染器
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Action");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Alert");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Remark");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/ButtonGroup");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/ButtonToolbar");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/DropDownButton");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Collapse");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Color");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/CRUD");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Pagination");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Cards");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Card");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Date");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Dialog");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Divider");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Each");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/index");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Control");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Hidden");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Text");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Tag");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Number");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Textarea");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Checkboxes");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Checkbox");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/City");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Rating");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Switch");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Button");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/ButtonGroup");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/ButtonToolbar");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Radios");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/List");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Location");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Select");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Static");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Date");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/DateRange");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Repeat");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Tree");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/TreeSelect");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Image");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/File");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Matrix");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Range");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Array");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Combo");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Container");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/SubForm");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/RichText");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Editor");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/DiffEditor");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Grid");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/HBox");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Panel");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Color");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/ChainedSelect");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/NestedSelect");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Transfer");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Service");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Table");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Picker");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/IconPicker");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Formula");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/FieldSet");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Tabs");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/TabsTransfer");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Group");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/InputGroup");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Grid");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/HBox");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/VBox");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Image");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Images");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/List");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Operation");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Page");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Panel");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Plain");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Spinner");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Table");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Tabs");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Tpl");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Mapping");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Progress");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Status");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Json");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Link");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Switch");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Wizard");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Chart");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Container");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Service");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Video");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Audio");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Nav");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Tasks");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Drawer");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Wrapper");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/IFrame");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/QRCode");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Icon");
  require("node_modules/amis-editor/node_modules/amis/lib/renderers/Carousel");
  var Scoped_1 = tslib_1.__importStar(require("node_modules/amis-editor/node_modules/amis/lib/Scoped"));
  exports.Scoped = Scoped_1.default;
  Object.defineProperty(exports, "ScopedContext", { enumerable: true, get: function () { return Scoped_1.ScopedContext; } });
  var Item_1 = require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Item");
  Object.defineProperty(exports, "FormItem", { enumerable: true, get: function () { return Item_1.FormItem; } });
  // 兼容旧版本用法
  require("node_modules/amis-editor/node_modules/amis/lib/compat");
  require("node_modules/amis-editor/node_modules/amis/lib/themes/default");
  require("node_modules/amis-editor/node_modules/amis/lib/themes/cxd");
  require("node_modules/amis-editor/node_modules/amis/lib/themes/dark");
  var tpl_builtin_1 = require("node_modules/amis-editor/node_modules/amis/lib/utils/tpl-builtin");
  Object.defineProperty(exports, "registerFilter", { enumerable: true, get: function () { return tpl_builtin_1.registerFilter; } });
  Object.defineProperty(exports, "filterDate", { enumerable: true, get: function () { return tpl_builtin_1.filterDate; } });
  Object.defineProperty(exports, "relativeValueRe", { enumerable: true, get: function () { return tpl_builtin_1.relativeValueRe; } });
  Object.defineProperty(exports, "resolveVariable", { enumerable: true, get: function () { return tpl_builtin_1.resolveVariable; } });
  Object.defineProperty(exports, "resolveVariableAndFilter", { enumerable: true, get: function () { return tpl_builtin_1.resolveVariableAndFilter; } });
  var validations_1 = require("node_modules/amis-editor/node_modules/amis/lib/utils/validations");
  Object.defineProperty(exports, "addRule", { enumerable: true, get: function () { return validations_1.addRule; } });
  Object.defineProperty(exports, "str2rules", { enumerable: true, get: function () { return validations_1.str2rules; } });
  Object.defineProperty(exports, "validate", { enumerable: true, get: function () { return validations_1.validate; } });
  Object.defineProperty(exports, "validateObject", { enumerable: true, get: function () { return validations_1.validateObject; } });
  var Select_1 = require("node_modules/amis-editor/node_modules/amis/lib/components/Select");
  Object.defineProperty(exports, "normalizeOptions", { enumerable: true, get: function () { return Select_1.normalizeOptions; } });
  var Options_1 = require("node_modules/amis-editor/node_modules/amis/lib/renderers/Form/Options");
  Object.defineProperty(exports, "OptionsControl", { enumerable: true, get: function () { return Options_1.OptionsControl; } });
  var theme_1 = require("node_modules/amis-editor/node_modules/amis/lib/theme");
  Object.defineProperty(exports, "classnames", { enumerable: true, get: function () { return theme_1.classnames; } });
  Object.defineProperty(exports, "getClassPrefix", { enumerable: true, get: function () { return theme_1.getClassPrefix; } });
  Object.defineProperty(exports, "setDefaultTheme", { enumerable: true, get: function () { return theme_1.setDefaultTheme; } });
  Object.defineProperty(exports, "theme", { enumerable: true, get: function () { return theme_1.theme; } });
  Object.defineProperty(exports, "getTheme", { enumerable: true, get: function () { return theme_1.getTheme; } });
  var classPrefix = theme_1.getClassPrefix();
  exports.classPrefix = classPrefix;
  tslib_1.__exportStar(require("node_modules/amis-editor/node_modules/amis/lib/components/index"), exports);
  

});
