// monaco-editor/esm/vs/platform/jsonschemas/common/jsonContributionRegistry.js
define('node_modules/monaco-editor/esm/vs/platform/jsonschemas/common/jsonContributionRegistry', function(require, exports, module) {

  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.Extensions = void 0;
  var tslib_1 = require("node_modules/tslib/tslib");
  /*---------------------------------------------------------------------------------------------
   *  Copyright (c) Microsoft Corporation. All rights reserved.
   *  Licensed under the MIT License. See License.txt in the project root for license information.
   *--------------------------------------------------------------------------------------------*/
  var platform = tslib_1.__importStar(require("node_modules/monaco-editor/esm/vs/platform/registry/common/platform"));
  var event_js_1 = require("node_modules/monaco-editor/esm/vs/base/common/event");
  exports.Extensions = {
      JSONContribution: 'base.contributions.json'
  };
  function normalizeId(id) {
      if (id.length > 0 && id.charAt(id.length - 1) === '#') {
          return id.substring(0, id.length - 1);
      }
      return id;
  }
  var JSONContributionRegistry = /** @class */ (function () {
      function JSONContributionRegistry() {
          this._onDidChangeSchema = new event_js_1.Emitter();
          this.schemasById = {};
      }
      JSONContributionRegistry.prototype.registerSchema = function (uri, unresolvedSchemaContent) {
          this.schemasById[normalizeId(uri)] = unresolvedSchemaContent;
          this._onDidChangeSchema.fire(uri);
      };
      JSONContributionRegistry.prototype.notifySchemaChanged = function (uri) {
          this._onDidChangeSchema.fire(uri);
      };
      return JSONContributionRegistry;
  }());
  var jsonContributionRegistry = new JSONContributionRegistry();
  platform.Registry.add(exports.Extensions.JSONContribution, jsonContributionRegistry);
  

});
