// react-color/lib/components/swatches/SwatchesGroup.js
define('node_modules/react-color/lib/components/swatches/SwatchesGroup', function(require, exports, module) {

  'use strict';
  
  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.SwatchesGroup = undefined;
  
  var _react = require('node_modules/react/index');
  
  var _react2 = _interopRequireDefault(_react);
  
  var _reactcss = require('node_modules/reactcss/lib/index');
  
  var _reactcss2 = _interopRequireDefault(_reactcss);
  
  var _map = require('node_modules/lodash/map');
  
  var _map2 = _interopRequireDefault(_map);
  
  var _SwatchesColor = require('node_modules/react-color/lib/components/swatches/SwatchesColor');
  
  var _SwatchesColor2 = _interopRequireDefault(_SwatchesColor);
  
  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
  
  var SwatchesGroup = exports.SwatchesGroup = function SwatchesGroup(_ref) {
    var onClick = _ref.onClick,
        onSwatchHover = _ref.onSwatchHover,
        group = _ref.group,
        active = _ref.active;
  
    var styles = (0, _reactcss2.default)({
      'default': {
        group: {
          paddingBottom: '10px',
          width: '40px',
          float: 'left',
          marginRight: '10px'
        }
      }
    });
  
    return _react2.default.createElement(
      'div',
      { style: styles.group },
      (0, _map2.default)(group, function (color, i) {
        return _react2.default.createElement(_SwatchesColor2.default, {
          key: color,
          color: color,
          active: color.toLowerCase() === active,
          first: i === 0,
          last: i === group.length - 1,
          onClick: onClick,
          onSwatchHover: onSwatchHover
        });
      })
    );
  };
  
  exports.default = SwatchesGroup;

});
