// monaco-editor/esm/vs/editor/contrib/toggleTabFocusMode/toggleTabFocusMode.js
define('node_modules/monaco-editor/esm/vs/editor/contrib/toggleTabFocusMode/toggleTabFocusMode', function(require, exports, module) {

  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.ToggleTabFocusModeAction = void 0;
  var tslib_1 = require("node_modules/tslib/tslib");
  /*---------------------------------------------------------------------------------------------
   *  Copyright (c) Microsoft Corporation. All rights reserved.
   *  Licensed under the MIT License. See License.txt in the project root for license information.
   *--------------------------------------------------------------------------------------------*/
  var __extends = (this && this.__extends) || (function () {
      var extendStatics = function (d, b) {
          extendStatics = Object.setPrototypeOf ||
              ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
              function (d, b) { for (var p in b)
                  if (b.hasOwnProperty(p))
                      d[p] = b[p]; };
          return extendStatics(d, b);
      };
      return function (d, b) {
          extendStatics(d, b);
          function __() { this.constructor = d; }
          d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
      };
  })();
  var nls = tslib_1.__importStar(require("node_modules/monaco-editor/esm/vs/nls"));
  var aria_js_1 = require("node_modules/monaco-editor/esm/vs/base/browser/ui/aria/aria");
  var editorExtensions_js_1 = require("node_modules/monaco-editor/esm/vs/editor/browser/editorExtensions");
  var commonEditorConfig_js_1 = require("node_modules/monaco-editor/esm/vs/editor/common/config/commonEditorConfig");
  var ToggleTabFocusModeAction = /** @class */ (function (_super) {
      __extends(ToggleTabFocusModeAction, _super);
      function ToggleTabFocusModeAction() {
          return _super.call(this, {
              id: ToggleTabFocusModeAction.ID,
              label: nls.localize({ key: 'toggle.tabMovesFocus', comment: ['Turn on/off use of tab key for moving focus around VS Code'] }, "Toggle Tab Key Moves Focus"),
              alias: 'Toggle Tab Key Moves Focus',
              precondition: null,
              kbOpts: {
                  kbExpr: null,
                  primary: 2048 /* CtrlCmd */ | 43 /* KEY_M */,
                  mac: { primary: 256 /* WinCtrl */ | 1024 /* Shift */ | 43 /* KEY_M */ },
                  weight: 100 /* EditorContrib */
              }
          }) || this;
      }
      ToggleTabFocusModeAction.prototype.run = function (accessor, editor) {
          var oldValue = commonEditorConfig_js_1.TabFocus.getTabFocusMode();
          var newValue = !oldValue;
          commonEditorConfig_js_1.TabFocus.setTabFocusMode(newValue);
          if (newValue) {
              aria_js_1.alert(nls.localize('toggle.tabMovesFocus.on', "Pressing Tab will now move focus to the next focusable element"));
          }
          else {
              aria_js_1.alert(nls.localize('toggle.tabMovesFocus.off', "Pressing Tab will now insert the tab character"));
          }
      };
      ToggleTabFocusModeAction.ID = 'editor.action.toggleTabFocusMode';
      return ToggleTabFocusModeAction;
  }(editorExtensions_js_1.EditorAction));
  exports.ToggleTabFocusModeAction = ToggleTabFocusModeAction;
  editorExtensions_js_1.registerEditorAction(ToggleTabFocusModeAction);
  

});
