// monaco-editor/esm/vs/editor/common/standalone/standaloneBase.js
define('node_modules/monaco-editor/esm/vs/editor/common/standalone/standaloneBase', function(require, exports, module) {

  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.createMonacoBaseAPI = exports.KeyMod = void 0;
  var tslib_1 = require("node_modules/tslib/tslib");
  /*---------------------------------------------------------------------------------------------
   *  Copyright (c) Microsoft Corporation. All rights reserved.
   *  Licensed under the MIT License. See License.txt in the project root for license information.
   *--------------------------------------------------------------------------------------------*/
  require("node_modules/monaco-editor/esm/vs/editor/common/standalone/promise-polyfill/polyfill");
  var cancellation_js_1 = require("node_modules/monaco-editor/esm/vs/base/common/cancellation");
  var event_js_1 = require("node_modules/monaco-editor/esm/vs/base/common/event");
  var keyCodes_js_1 = require("node_modules/monaco-editor/esm/vs/base/common/keyCodes");
  var uri_js_1 = require("node_modules/monaco-editor/esm/vs/base/common/uri");
  var position_js_1 = require("node_modules/monaco-editor/esm/vs/editor/common/core/position");
  var range_js_1 = require("node_modules/monaco-editor/esm/vs/editor/common/core/range");
  var selection_js_1 = require("node_modules/monaco-editor/esm/vs/editor/common/core/selection");
  var token_js_1 = require("node_modules/monaco-editor/esm/vs/editor/common/core/token");
  var standaloneEnums = tslib_1.__importStar(require("node_modules/monaco-editor/esm/vs/editor/common/standalone/standaloneEnums"));
  var KeyMod = /** @class */ (function () {
      function KeyMod() {
      }
      KeyMod.chord = function (firstPart, secondPart) {
          return keyCodes_js_1.KeyChord(firstPart, secondPart);
      };
      KeyMod.CtrlCmd = 2048 /* CtrlCmd */;
      KeyMod.Shift = 1024 /* Shift */;
      KeyMod.Alt = 512 /* Alt */;
      KeyMod.WinCtrl = 256 /* WinCtrl */;
      return KeyMod;
  }());
  exports.KeyMod = KeyMod;
  function createMonacoBaseAPI() {
      return {
          editor: undefined,
          languages: undefined,
          CancellationTokenSource: cancellation_js_1.CancellationTokenSource,
          Emitter: event_js_1.Emitter,
          KeyCode: standaloneEnums.KeyCode,
          KeyMod: KeyMod,
          Position: position_js_1.Position,
          Range: range_js_1.Range,
          Selection: selection_js_1.Selection,
          SelectionDirection: standaloneEnums.SelectionDirection,
          MarkerSeverity: standaloneEnums.MarkerSeverity,
          MarkerTag: standaloneEnums.MarkerTag,
          Uri: uri_js_1.URI,
          Token: token_js_1.Token
      };
  }
  exports.createMonacoBaseAPI = createMonacoBaseAPI;
  

});
