// react-color/lib/index.js
define('node_modules/react-color/lib/index', function(require, exports, module) {

  'use strict';
  
  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.CustomPicker = exports.TwitterPicker = exports.SwatchesPicker = exports.SliderPicker = exports.SketchPicker = exports.PhotoshopPicker = exports.MaterialPicker = exports.HuePicker = exports.GithubPicker = exports.CompactPicker = exports.ChromePicker = exports.default = exports.CirclePicker = exports.BlockPicker = exports.AlphaPicker = undefined;
  
  var _Alpha = require('node_modules/react-color/lib/components/alpha/Alpha');
  
  Object.defineProperty(exports, 'AlphaPicker', {
    enumerable: true,
    get: function get() {
      return _interopRequireDefault(_Alpha).default;
    }
  });
  
  var _Block = require('node_modules/react-color/lib/components/block/Block');
  
  Object.defineProperty(exports, 'BlockPicker', {
    enumerable: true,
    get: function get() {
      return _interopRequireDefault(_Block).default;
    }
  });
  
  var _Circle = require('node_modules/react-color/lib/components/circle/Circle');
  
  Object.defineProperty(exports, 'CirclePicker', {
    enumerable: true,
    get: function get() {
      return _interopRequireDefault(_Circle).default;
    }
  });
  
  var _Chrome = require('node_modules/react-color/lib/components/chrome/Chrome');
  
  Object.defineProperty(exports, 'ChromePicker', {
    enumerable: true,
    get: function get() {
      return _interopRequireDefault(_Chrome).default;
    }
  });
  
  var _Compact = require('node_modules/react-color/lib/components/compact/Compact');
  
  Object.defineProperty(exports, 'CompactPicker', {
    enumerable: true,
    get: function get() {
      return _interopRequireDefault(_Compact).default;
    }
  });
  
  var _Github = require('node_modules/react-color/lib/components/github/Github');
  
  Object.defineProperty(exports, 'GithubPicker', {
    enumerable: true,
    get: function get() {
      return _interopRequireDefault(_Github).default;
    }
  });
  
  var _Hue = require('node_modules/react-color/lib/components/hue/Hue');
  
  Object.defineProperty(exports, 'HuePicker', {
    enumerable: true,
    get: function get() {
      return _interopRequireDefault(_Hue).default;
    }
  });
  
  var _Material = require('node_modules/react-color/lib/components/material/Material');
  
  Object.defineProperty(exports, 'MaterialPicker', {
    enumerable: true,
    get: function get() {
      return _interopRequireDefault(_Material).default;
    }
  });
  
  var _Photoshop = require('node_modules/react-color/lib/components/photoshop/Photoshop');
  
  Object.defineProperty(exports, 'PhotoshopPicker', {
    enumerable: true,
    get: function get() {
      return _interopRequireDefault(_Photoshop).default;
    }
  });
  
  var _Sketch = require('node_modules/react-color/lib/components/sketch/Sketch');
  
  Object.defineProperty(exports, 'SketchPicker', {
    enumerable: true,
    get: function get() {
      return _interopRequireDefault(_Sketch).default;
    }
  });
  
  var _Slider = require('node_modules/react-color/lib/components/slider/Slider');
  
  Object.defineProperty(exports, 'SliderPicker', {
    enumerable: true,
    get: function get() {
      return _interopRequireDefault(_Slider).default;
    }
  });
  
  var _Swatches = require('node_modules/react-color/lib/components/swatches/Swatches');
  
  Object.defineProperty(exports, 'SwatchesPicker', {
    enumerable: true,
    get: function get() {
      return _interopRequireDefault(_Swatches).default;
    }
  });
  
  var _Twitter = require('node_modules/react-color/lib/components/twitter/Twitter');
  
  Object.defineProperty(exports, 'TwitterPicker', {
    enumerable: true,
    get: function get() {
      return _interopRequireDefault(_Twitter).default;
    }
  });
  
  var _ColorWrap = require('node_modules/react-color/lib/components/common/ColorWrap');
  
  Object.defineProperty(exports, 'CustomPicker', {
    enumerable: true,
    get: function get() {
      return _interopRequireDefault(_ColorWrap).default;
    }
  });
  
  var _Chrome2 = _interopRequireDefault(_Chrome);
  
  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
  
  exports.default = _Chrome2.default;

});
