// monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/cssLanguageService.js
define('node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/cssLanguageService', function(require, exports, module) {

  /*---------------------------------------------------------------------------------------------
   *  Copyright (c) Microsoft Corporation. All rights reserved.
   *  Licensed under the MIT License. See License.txt in the project root for license information.
   *--------------------------------------------------------------------------------------------*/
  'use strict';
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.getLESSLanguageService = exports.getSCSSLanguageService = exports.getCSSLanguageService = void 0;
  var tslib_1 = require("node_modules/tslib/tslib");
  var cssParser_js_1 = require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/parser/cssParser");
  var cssCompletion_js_1 = require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/services/cssCompletion");
  var cssHover_js_1 = require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/services/cssHover");
  var cssNavigation_js_1 = require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/services/cssNavigation");
  var cssCodeActions_js_1 = require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/services/cssCodeActions");
  var cssValidation_js_1 = require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/services/cssValidation");
  var scssParser_js_1 = require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/parser/scssParser");
  var scssCompletion_js_1 = require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/services/scssCompletion");
  var lessParser_js_1 = require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/parser/lessParser");
  var lessCompletion_js_1 = require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/services/lessCompletion");
  var cssFolding_js_1 = require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/services/cssFolding");
  var facts_js_1 = require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/languageFacts/facts");
  var cssSelectionRange_js_1 = require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/services/cssSelectionRange");
  tslib_1.__exportStar(require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-css-languageservice/cssLanguageTypes"), exports);
  tslib_1.__exportStar(require("node_modules/monaco-editor/esm/vs/language/css/_deps/vscode-languageserver-types/main"), exports);
  function createFacade(parser, completion, hover, navigation, codeActions, validation) {
      return {
          configure: validation.configure.bind(validation),
          doValidation: validation.doValidation.bind(validation),
          parseStylesheet: parser.parseStylesheet.bind(parser),
          doComplete: completion.doComplete.bind(completion),
          setCompletionParticipants: completion.setCompletionParticipants.bind(completion),
          doHover: hover.doHover.bind(hover),
          findDefinition: navigation.findDefinition.bind(navigation),
          findReferences: navigation.findReferences.bind(navigation),
          findDocumentHighlights: navigation.findDocumentHighlights.bind(navigation),
          findDocumentLinks: navigation.findDocumentLinks.bind(navigation),
          findDocumentSymbols: navigation.findDocumentSymbols.bind(navigation),
          doCodeActions: codeActions.doCodeActions.bind(codeActions),
          doCodeActions2: codeActions.doCodeActions2.bind(codeActions),
          findColorSymbols: function (d, s) { return navigation.findDocumentColors(d, s).map(function (s) { return s.range; }); },
          findDocumentColors: navigation.findDocumentColors.bind(navigation),
          getColorPresentations: navigation.getColorPresentations.bind(navigation),
          doRename: navigation.doRename.bind(navigation),
          getFoldingRanges: cssFolding_js_1.getFoldingRanges,
          getSelectionRanges: cssSelectionRange_js_1.getSelectionRanges
      };
  }
  function handleCustomData(options) {
      if (options && options.customDataProviders) {
          facts_js_1.cssDataManager.addDataProviders(options.customDataProviders);
      }
  }
  function getCSSLanguageService(options) {
      handleCustomData(options);
      return createFacade(new cssParser_js_1.Parser(), new cssCompletion_js_1.CSSCompletion(), new cssHover_js_1.CSSHover(), new cssNavigation_js_1.CSSNavigation(), new cssCodeActions_js_1.CSSCodeActions(), new cssValidation_js_1.CSSValidation());
  }
  exports.getCSSLanguageService = getCSSLanguageService;
  function getSCSSLanguageService(options) {
      handleCustomData(options);
      return createFacade(new scssParser_js_1.SCSSParser(), new scssCompletion_js_1.SCSSCompletion(), new cssHover_js_1.CSSHover(), new cssNavigation_js_1.CSSNavigation(), new cssCodeActions_js_1.CSSCodeActions(), new cssValidation_js_1.CSSValidation());
  }
  exports.getSCSSLanguageService = getSCSSLanguageService;
  function getLESSLanguageService(options) {
      handleCustomData(options);
      return createFacade(new lessParser_js_1.LESSParser(), new lessCompletion_js_1.LESSCompletion(), new cssHover_js_1.CSSHover(), new cssNavigation_js_1.CSSNavigation(), new cssCodeActions_js_1.CSSCodeActions(), new cssValidation_js_1.CSSValidation());
  }
  exports.getLESSLanguageService = getLESSLanguageService;
  

});
