// monaco-editor/esm/vs/editor/editor.api.js
define('node_modules/monaco-editor/esm/vs/editor/editor.api', function(require, exports, module) {

  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.languages = exports.editor = exports.Token = exports.Uri = exports.MarkerTag = exports.MarkerSeverity = exports.SelectionDirection = exports.Selection = exports.Range = exports.Position = exports.KeyMod = exports.KeyCode = exports.Emitter = exports.CancellationTokenSource = void 0;
  /*---------------------------------------------------------------------------------------------
   *  Copyright (c) Microsoft Corporation. All rights reserved.
   *  Licensed under the MIT License. See License.txt in the project root for license information.
   *--------------------------------------------------------------------------------------------*/
  var editorOptions_js_1 = require("node_modules/monaco-editor/esm/vs/editor/common/config/editorOptions");
  var standaloneBase_js_1 = require("node_modules/monaco-editor/esm/vs/editor/common/standalone/standaloneBase");
  var standaloneEditor_js_1 = require("node_modules/monaco-editor/esm/vs/editor/standalone/browser/standaloneEditor");
  var standaloneLanguages_js_1 = require("node_modules/monaco-editor/esm/vs/editor/standalone/browser/standaloneLanguages");
  var global = self;
  // Set defaults for standalone editor
  editorOptions_js_1.EDITOR_DEFAULTS.wrappingIndent = 0 /* None */;
  editorOptions_js_1.EDITOR_DEFAULTS.viewInfo.glyphMargin = false;
  editorOptions_js_1.EDITOR_DEFAULTS.autoIndent = false;
  var api = standaloneBase_js_1.createMonacoBaseAPI();
  api.editor = standaloneEditor_js_1.createMonacoEditorAPI();
  api.languages = standaloneLanguages_js_1.createMonacoLanguagesAPI();
  exports.CancellationTokenSource = api.CancellationTokenSource;
  exports.Emitter = api.Emitter;
  exports.KeyCode = api.KeyCode;
  exports.KeyMod = api.KeyMod;
  exports.Position = api.Position;
  exports.Range = api.Range;
  exports.Selection = api.Selection;
  exports.SelectionDirection = api.SelectionDirection;
  exports.MarkerSeverity = api.MarkerSeverity;
  exports.MarkerTag = api.MarkerTag;
  exports.Uri = api.Uri;
  exports.Token = api.Token;
  exports.editor = api.editor;
  exports.languages = api.languages;
  global.monaco = api;
  if (typeof global.require !== 'undefined' && typeof global.require.config === 'function') {
      global.require.config({
          ignoreDuplicateModules: [
              'vscode-languageserver-types',
              'vscode-languageserver-types/main',
              'vscode-nls',
              'vscode-nls/vscode-nls',
              'jsonc-parser',
              'jsonc-parser/main',
              'vscode-uri',
              'vscode-uri/index',
              'vs/basic-languages/typescript/typescript'
          ]
      });
  }
  

});
