// monaco-editor/esm/vs/editor/contrib/referenceSearch/referencesTree.js
define('node_modules/monaco-editor/esm/vs/editor/contrib/referenceSearch/referencesTree', function(require, exports, module) {

  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.AriaProvider = exports.OneReferenceRenderer = exports.FileReferencesRenderer = exports.IdentityProvider = exports.StringRepresentationProvider = exports.Delegate = exports.DataSource = void 0;
  var tslib_1 = require("node_modules/tslib/tslib");
  /*---------------------------------------------------------------------------------------------
   *  Copyright (c) Microsoft Corporation. All rights reserved.
   *  Licensed under the MIT License. See License.txt in the project root for license information.
   *--------------------------------------------------------------------------------------------*/
  var __extends = (this && this.__extends) || (function () {
      var extendStatics = function (d, b) {
          extendStatics = Object.setPrototypeOf ||
              ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
              function (d, b) { for (var p in b)
                  if (b.hasOwnProperty(p))
                      d[p] = b[p]; };
          return extendStatics(d, b);
      };
      return function (d, b) {
          extendStatics(d, b);
          function __() { this.constructor = d; }
          d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
      };
  })();
  var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
      var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
      if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
          r = Reflect.decorate(decorators, target, key, desc);
      else
          for (var i = decorators.length - 1; i >= 0; i--)
              if (d = decorators[i])
                  r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
      return c > 3 && r && Object.defineProperty(target, key, r), r;
  };
  var __param = (this && this.__param) || function (paramIndex, decorator) {
      return function (target, key) { decorator(target, key, paramIndex); };
  };
  var referencesModel_js_1 = require("node_modules/monaco-editor/esm/vs/editor/contrib/referenceSearch/referencesModel");
  var resolverService_js_1 = require("node_modules/monaco-editor/esm/vs/editor/common/services/resolverService");
  var iconLabel_js_1 = require("node_modules/monaco-editor/esm/vs/base/browser/ui/iconLabel/iconLabel");
  var countBadge_js_1 = require("node_modules/monaco-editor/esm/vs/base/browser/ui/countBadge/countBadge");
  var label_js_1 = require("node_modules/monaco-editor/esm/vs/platform/label/common/label");
  var themeService_js_1 = require("node_modules/monaco-editor/esm/vs/platform/theme/common/themeService");
  var styler_js_1 = require("node_modules/monaco-editor/esm/vs/platform/theme/common/styler");
  var dom = tslib_1.__importStar(require("node_modules/monaco-editor/esm/vs/base/browser/dom"));
  var nls_js_1 = require("node_modules/monaco-editor/esm/vs/nls");
  var labels_js_1 = require("node_modules/monaco-editor/esm/vs/base/common/labels");
  var resources_js_1 = require("node_modules/monaco-editor/esm/vs/base/common/resources");
  var strings_js_1 = require("node_modules/monaco-editor/esm/vs/base/common/strings");
  var lifecycle_js_1 = require("node_modules/monaco-editor/esm/vs/base/common/lifecycle");
  var instantiation_js_1 = require("node_modules/monaco-editor/esm/vs/platform/instantiation/common/instantiation");
  var keybinding_js_1 = require("node_modules/monaco-editor/esm/vs/platform/keybinding/common/keybinding");
  var filters_js_1 = require("node_modules/monaco-editor/esm/vs/base/common/filters");
  var DataSource = /** @class */ (function () {
      function DataSource(_resolverService) {
          this._resolverService = _resolverService;
      }
      DataSource.prototype.hasChildren = function (element) {
          if (element instanceof referencesModel_js_1.ReferencesModel) {
              return true;
          }
          if (element instanceof referencesModel_js_1.FileReferences && !element.failure) {
              return true;
          }
          return false;
      };
      DataSource.prototype.getChildren = function (element) {
          if (element instanceof referencesModel_js_1.ReferencesModel) {
              return element.groups;
          }
          if (element instanceof referencesModel_js_1.FileReferences) {
              return element.resolve(this._resolverService).then(function (val) {
                  // if (element.failure) {
                  // 	// refresh the element on failure so that
                  // 	// we can update its rendering
                  // 	return tree.refresh(element).then(() => val.children);
                  // }
                  return val.children;
              });
          }
          throw new Error('bad tree');
      };
      DataSource = __decorate([
          __param(0, resolverService_js_1.ITextModelService)
      ], DataSource);
      return DataSource;
  }());
  exports.DataSource = DataSource;
  //#endregion
  var Delegate = /** @class */ (function () {
      function Delegate() {
      }
      Delegate.prototype.getHeight = function () {
          return 23;
      };
      Delegate.prototype.getTemplateId = function (element) {
          if (element instanceof referencesModel_js_1.FileReferences) {
              return FileReferencesRenderer.id;
          }
          else {
              return OneReferenceRenderer.id;
          }
      };
      return Delegate;
  }());
  exports.Delegate = Delegate;
  var StringRepresentationProvider = /** @class */ (function () {
      function StringRepresentationProvider(_keybindingService) {
          this._keybindingService = _keybindingService;
      }
      StringRepresentationProvider.prototype.getKeyboardNavigationLabel = function (element) {
          // todo@joao `OneReference` elements are lazy and their "real" label
          // isn't known yet
          return resources_js_1.basename(element.uri);
      };
      StringRepresentationProvider.prototype.mightProducePrintableCharacter = function (event) {
          return this._keybindingService.mightProducePrintableCharacter(event);
      };
      StringRepresentationProvider = __decorate([
          __param(0, keybinding_js_1.IKeybindingService)
      ], StringRepresentationProvider);
      return StringRepresentationProvider;
  }());
  exports.StringRepresentationProvider = StringRepresentationProvider;
  var IdentityProvider = /** @class */ (function () {
      function IdentityProvider() {
      }
      IdentityProvider.prototype.getId = function (element) {
          return element.id;
      };
      return IdentityProvider;
  }());
  exports.IdentityProvider = IdentityProvider;
  //#region render: File
  var FileReferencesTemplate = /** @class */ (function (_super) {
      __extends(FileReferencesTemplate, _super);
      function FileReferencesTemplate(container, _uriLabel, themeService) {
          var _this = _super.call(this) || this;
          _this._uriLabel = _uriLabel;
          var parent = document.createElement('div');
          dom.addClass(parent, 'reference-file');
          _this.file = _this._register(new iconLabel_js_1.IconLabel(parent, { supportHighlights: true }));
          _this.badge = new countBadge_js_1.CountBadge(dom.append(parent, dom.$('.count')));
          _this._register(styler_js_1.attachBadgeStyler(_this.badge, themeService));
          container.appendChild(parent);
          return _this;
      }
      FileReferencesTemplate.prototype.set = function (element, matches) {
          var parent = resources_js_1.dirname(element.uri);
          this.file.setLabel(labels_js_1.getBaseLabel(element.uri), this._uriLabel.getUriLabel(parent, { relative: true }), { title: this._uriLabel.getUriLabel(element.uri), matches: matches });
          var len = element.children.length;
          this.badge.setCount(len);
          if (element.failure) {
              this.badge.setTitleFormat(nls_js_1.localize('referencesFailre', "Failed to resolve file."));
          }
          else if (len > 1) {
              this.badge.setTitleFormat(nls_js_1.localize('referencesCount', "{0} references", len));
          }
          else {
              this.badge.setTitleFormat(nls_js_1.localize('referenceCount', "{0} reference", len));
          }
      };
      FileReferencesTemplate = __decorate([
          __param(1, label_js_1.ILabelService),
          __param(2, themeService_js_1.IThemeService)
      ], FileReferencesTemplate);
      return FileReferencesTemplate;
  }(lifecycle_js_1.Disposable));
  var FileReferencesRenderer = /** @class */ (function () {
      function FileReferencesRenderer(_instantiationService) {
          this._instantiationService = _instantiationService;
          this.templateId = FileReferencesRenderer.id;
      }
      FileReferencesRenderer.prototype.renderTemplate = function (container) {
          return this._instantiationService.createInstance(FileReferencesTemplate, container);
      };
      FileReferencesRenderer.prototype.renderElement = function (node, index, template) {
          template.set(node.element, filters_js_1.createMatches(node.filterData));
      };
      FileReferencesRenderer.prototype.disposeTemplate = function (templateData) {
          templateData.dispose();
      };
      FileReferencesRenderer.id = 'FileReferencesRenderer';
      FileReferencesRenderer = __decorate([
          __param(0, instantiation_js_1.IInstantiationService)
      ], FileReferencesRenderer);
      return FileReferencesRenderer;
  }());
  exports.FileReferencesRenderer = FileReferencesRenderer;
  //#endregion
  //#region render: Reference
  var OneReferenceTemplate = /** @class */ (function () {
      function OneReferenceTemplate(container) {
          var parent = document.createElement('div');
          this.before = document.createElement('span');
          this.inside = document.createElement('span');
          this.after = document.createElement('span');
          dom.addClass(this.inside, 'referenceMatch');
          dom.addClass(parent, 'reference');
          parent.appendChild(this.before);
          parent.appendChild(this.inside);
          parent.appendChild(this.after);
          container.appendChild(parent);
      }
      OneReferenceTemplate.prototype.set = function (element) {
          var filePreview = element.parent.preview;
          var preview = filePreview && filePreview.preview(element.range);
          if (preview) {
              var before = preview.before, inside = preview.inside, after = preview.after;
              this.before.innerHTML = strings_js_1.escape(before);
              this.inside.innerHTML = strings_js_1.escape(inside);
              this.after.innerHTML = strings_js_1.escape(after);
          }
      };
      return OneReferenceTemplate;
  }());
  var OneReferenceRenderer = /** @class */ (function () {
      function OneReferenceRenderer() {
          this.templateId = OneReferenceRenderer.id;
      }
      OneReferenceRenderer.prototype.renderTemplate = function (container) {
          return new OneReferenceTemplate(container);
      };
      OneReferenceRenderer.prototype.renderElement = function (element, index, templateData) {
          templateData.set(element.element);
      };
      OneReferenceRenderer.prototype.disposeTemplate = function () {
          //
      };
      OneReferenceRenderer.id = 'OneReferenceRenderer';
      return OneReferenceRenderer;
  }());
  exports.OneReferenceRenderer = OneReferenceRenderer;
  //#endregion
  var AriaProvider = /** @class */ (function () {
      function AriaProvider() {
      }
      AriaProvider.prototype.getAriaLabel = function (element) {
          if (element instanceof referencesModel_js_1.FileReferences) {
              return element.getAriaMessage();
          }
          else if (element instanceof referencesModel_js_1.OneReference) {
              return element.getAriaMessage();
          }
          else {
              return null;
          }
      };
      return AriaProvider;
  }());
  exports.AriaProvider = AriaProvider;
  

});
