// amis/lib/components/SearchBox.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var theme_1 = require("../theme");
var icons_1 = require("./icons");
var uncontrollable_1 = require("uncontrollable");
var helper_1 = require("../utils/helper");
var locale_1 = require("../locale");
var SearchBox = /** @class */ (function (_super) {
    tslib_1.__extends(SearchBox, _super);
    function SearchBox() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.inputRef = react_1.default.createRef();
        return _this;
    }
    SearchBox.prototype.handleActive = function () {
        var _a, _b;
        var onActiveChange = this.props.onActiveChange;
        (_a = onActiveChange) === null || _a === void 0 ? void 0 : _a(true);
        (_b = this.inputRef.current) === null || _b === void 0 ? void 0 : _b.focus();
    };
    SearchBox.prototype.handleCancel = function () {
        var _a, _b, _c;
        var _d = this.props, onActiveChange = _d.onActiveChange, onChange = _d.onChange, onCancel = _d.onCancel;
        (_a = onActiveChange) === null || _a === void 0 ? void 0 : _a(false);
        (_b = onCancel) === null || _b === void 0 ? void 0 : _b();
        (_c = onChange) === null || _c === void 0 ? void 0 : _c('');
    };
    SearchBox.prototype.handleChange = function (e) {
        var _a, _b;
        var _c = this.props, onChange = _c.onChange, onSearch = _c.onSearch;
        (_a = onChange) === null || _a === void 0 ? void 0 : _a(e.currentTarget.value);
        (_b = onSearch) === null || _b === void 0 ? void 0 : _b(e.currentTarget.value);
    };
    SearchBox.prototype.render = function () {
        var _a = this.props, cx = _a.classnames, value = _a.value, active = _a.active, name = _a.name, onChange = _a.onChange, placeholder = _a.placeholder, __ = _a.translate;
        return (react_1.default.createElement("div", { className: cx('SearchBox', active ? 'is-active' : '') },
            react_1.default.createElement("input", { name: name, onChange: this.handleChange, value: value || '', placeholder: __(placeholder || '输入关键字'), ref: this.inputRef }),
            active ? (react_1.default.createElement("a", { className: cx('SearchBox-cancelBtn'), onClick: this.handleCancel },
                react_1.default.createElement(icons_1.Icon, { icon: "close", className: "icon" }))) : (react_1.default.createElement("a", { className: cx('SearchBox-activeBtn'), onClick: this.handleActive },
                react_1.default.createElement(icons_1.Icon, { icon: "search", className: "icon" })))));
    };
    var _a;
    tslib_1.__decorate([
        helper_1.autobind,
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], SearchBox.prototype, "handleActive", null);
    tslib_1.__decorate([
        helper_1.autobind,
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], SearchBox.prototype, "handleCancel", null);
    tslib_1.__decorate([
        helper_1.autobind,
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", [typeof (_a = typeof react_1.default !== "undefined" && react_1.default.ChangeEvent) === "function" ? _a : Object]),
        tslib_1.__metadata("design:returntype", void 0)
    ], SearchBox.prototype, "handleChange", null);
    return SearchBox;
}(react_1.default.Component));
exports.SearchBox = SearchBox;
exports.default = theme_1.themeable(locale_1.localeable(uncontrollable_1.uncontrollable(SearchBox, {
    active: 'onActiveChange',
    value: 'onChange'
})));
