// monaco-editor/esm/vs/editor/contrib/parameterHints/parameterHints.js
define('node_modules/monaco-editor/esm/vs/editor/contrib/parameterHints/parameterHints', function(require, exports, module) {

  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.TriggerParameterHintsAction = void 0;
  var tslib_1 = require("node_modules/tslib/tslib");
  /*---------------------------------------------------------------------------------------------
   *  Copyright (c) Microsoft Corporation. All rights reserved.
   *  Licensed under the MIT License. See License.txt in the project root for license information.
   *--------------------------------------------------------------------------------------------*/
  var __extends = (this && this.__extends) || (function () {
      var extendStatics = function (d, b) {
          extendStatics = Object.setPrototypeOf ||
              ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
              function (d, b) { for (var p in b)
                  if (b.hasOwnProperty(p))
                      d[p] = b[p]; };
          return extendStatics(d, b);
      };
      return function (d, b) {
          extendStatics(d, b);
          function __() { this.constructor = d; }
          d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
      };
  })();
  var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
      var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
      if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
          r = Reflect.decorate(decorators, target, key, desc);
      else
          for (var i = decorators.length - 1; i >= 0; i--)
              if (d = decorators[i])
                  r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
      return c > 3 && r && Object.defineProperty(target, key, r), r;
  };
  var __param = (this && this.__param) || function (paramIndex, decorator) {
      return function (target, key) { decorator(target, key, paramIndex); };
  };
  var nls = tslib_1.__importStar(require("node_modules/monaco-editor/esm/vs/nls"));
  var lifecycle_js_1 = require("node_modules/monaco-editor/esm/vs/base/common/lifecycle");
  var instantiation_js_1 = require("node_modules/monaco-editor/esm/vs/platform/instantiation/common/instantiation");
  var editorContextKeys_js_1 = require("node_modules/monaco-editor/esm/vs/editor/common/editorContextKeys");
  var contextkey_js_1 = require("node_modules/monaco-editor/esm/vs/platform/contextkey/common/contextkey");
  var editorExtensions_js_1 = require("node_modules/monaco-editor/esm/vs/editor/browser/editorExtensions");
  var parameterHintsWidget_js_1 = require("node_modules/monaco-editor/esm/vs/editor/contrib/parameterHints/parameterHintsWidget");
  var provideSignatureHelp_js_1 = require("node_modules/monaco-editor/esm/vs/editor/contrib/parameterHints/provideSignatureHelp");
  var modes = tslib_1.__importStar(require("node_modules/monaco-editor/esm/vs/editor/common/modes"));
  var ParameterHintsController = /** @class */ (function () {
      function ParameterHintsController(editor, instantiationService) {
          this.editor = editor;
          this.widget = instantiationService.createInstance(parameterHintsWidget_js_1.ParameterHintsWidget, this.editor);
      }
      ParameterHintsController.get = function (editor) {
          return editor.getContribution(ParameterHintsController.ID);
      };
      ParameterHintsController.prototype.getId = function () {
          return ParameterHintsController.ID;
      };
      ParameterHintsController.prototype.cancel = function () {
          this.widget.cancel();
      };
      ParameterHintsController.prototype.previous = function () {
          this.widget.previous();
      };
      ParameterHintsController.prototype.next = function () {
          this.widget.next();
      };
      ParameterHintsController.prototype.trigger = function (context) {
          this.widget.trigger(context);
      };
      ParameterHintsController.prototype.dispose = function () {
          lifecycle_js_1.dispose(this.widget);
      };
      ParameterHintsController.ID = 'editor.controller.parameterHints';
      ParameterHintsController = __decorate([
          __param(1, instantiation_js_1.IInstantiationService)
      ], ParameterHintsController);
      return ParameterHintsController;
  }());
  var TriggerParameterHintsAction = /** @class */ (function (_super) {
      __extends(TriggerParameterHintsAction, _super);
      function TriggerParameterHintsAction() {
          return _super.call(this, {
              id: 'editor.action.triggerParameterHints',
              label: nls.localize('parameterHints.trigger.label', "Trigger Parameter Hints"),
              alias: 'Trigger Parameter Hints',
              precondition: editorContextKeys_js_1.EditorContextKeys.hasSignatureHelpProvider,
              kbOpts: {
                  kbExpr: editorContextKeys_js_1.EditorContextKeys.editorTextFocus,
                  primary: 2048 /* CtrlCmd */ | 1024 /* Shift */ | 10 /* Space */,
                  weight: 100 /* EditorContrib */
              }
          }) || this;
      }
      TriggerParameterHintsAction.prototype.run = function (accessor, editor) {
          var controller = ParameterHintsController.get(editor);
          if (controller) {
              controller.trigger({
                  triggerKind: modes.SignatureHelpTriggerKind.Invoke
              });
          }
      };
      return TriggerParameterHintsAction;
  }(editorExtensions_js_1.EditorAction));
  exports.TriggerParameterHintsAction = TriggerParameterHintsAction;
  editorExtensions_js_1.registerEditorContribution(ParameterHintsController);
  editorExtensions_js_1.registerEditorAction(TriggerParameterHintsAction);
  var weight = 100 /* EditorContrib */ + 75;
  var ParameterHintsCommand = editorExtensions_js_1.EditorCommand.bindToContribution(ParameterHintsController.get);
  editorExtensions_js_1.registerEditorCommand(new ParameterHintsCommand({
      id: 'closeParameterHints',
      precondition: provideSignatureHelp_js_1.Context.Visible,
      handler: function (x) { return x.cancel(); },
      kbOpts: {
          weight: weight,
          kbExpr: editorContextKeys_js_1.EditorContextKeys.focus,
          primary: 9 /* Escape */,
          secondary: [1024 /* Shift */ | 9 /* Escape */]
      }
  }));
  editorExtensions_js_1.registerEditorCommand(new ParameterHintsCommand({
      id: 'showPrevParameterHint',
      precondition: contextkey_js_1.ContextKeyExpr.and(provideSignatureHelp_js_1.Context.Visible, provideSignatureHelp_js_1.Context.MultipleSignatures),
      handler: function (x) { return x.previous(); },
      kbOpts: {
          weight: weight,
          kbExpr: editorContextKeys_js_1.EditorContextKeys.focus,
          primary: 16 /* UpArrow */,
          secondary: [512 /* Alt */ | 16 /* UpArrow */],
          mac: { primary: 16 /* UpArrow */, secondary: [512 /* Alt */ | 16 /* UpArrow */, 256 /* WinCtrl */ | 46 /* KEY_P */] }
      }
  }));
  editorExtensions_js_1.registerEditorCommand(new ParameterHintsCommand({
      id: 'showNextParameterHint',
      precondition: contextkey_js_1.ContextKeyExpr.and(provideSignatureHelp_js_1.Context.Visible, provideSignatureHelp_js_1.Context.MultipleSignatures),
      handler: function (x) { return x.next(); },
      kbOpts: {
          weight: weight,
          kbExpr: editorContextKeys_js_1.EditorContextKeys.focus,
          primary: 18 /* DownArrow */,
          secondary: [512 /* Alt */ | 18 /* DownArrow */],
          mac: { primary: 18 /* DownArrow */, secondary: [512 /* Alt */ | 18 /* DownArrow */, 256 /* WinCtrl */ | 44 /* KEY_N */] }
      }
  }));
  

});
