// react-color/lib/components/compact/Compact.js
define('node_modules/react-color/lib/components/compact/Compact', function(require, exports, module) {

  'use strict';
  
  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.Compact = undefined;
  
  var _react = require('node_modules/react/index');
  
  var _react2 = _interopRequireDefault(_react);
  
  var _propTypes = require('node_modules/prop-types/index');
  
  var _propTypes2 = _interopRequireDefault(_propTypes);
  
  var _reactcss = require('node_modules/reactcss/lib/index');
  
  var _reactcss2 = _interopRequireDefault(_reactcss);
  
  var _map = require('node_modules/lodash/map');
  
  var _map2 = _interopRequireDefault(_map);
  
  var _color = require('node_modules/react-color/lib/helpers/color');
  
  var _color2 = _interopRequireDefault(_color);
  
  var _common = require('node_modules/react-color/lib/components/common/index');
  
  var _CompactColor = require('node_modules/react-color/lib/components/compact/CompactColor');
  
  var _CompactColor2 = _interopRequireDefault(_CompactColor);
  
  var _CompactFields = require('node_modules/react-color/lib/components/compact/CompactFields');
  
  var _CompactFields2 = _interopRequireDefault(_CompactFields);
  
  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
  
  var Compact = exports.Compact = function Compact(_ref) {
    var onChange = _ref.onChange,
        onSwatchHover = _ref.onSwatchHover,
        colors = _ref.colors,
        hex = _ref.hex,
        rgb = _ref.rgb,
        _ref$className = _ref.className,
        className = _ref$className === undefined ? '' : _ref$className;
  
    var styles = (0, _reactcss2.default)({
      'default': {
        Compact: {
          background: '#f6f6f6',
          radius: '4px'
        },
        compact: {
          paddingTop: '5px',
          paddingLeft: '5px',
          boxSizing: 'initial',
          width: '240px'
        },
        clear: {
          clear: 'both'
        }
      }
    });
  
    var handleChange = function handleChange(data, e) {
      if (data.hex) {
        _color2.default.isValidHex(data.hex) && onChange({
          hex: data.hex,
          source: 'hex'
        }, e);
      } else {
        onChange(data, e);
      }
    };
  
    return _react2.default.createElement(
      _common.Raised,
      { style: styles.Compact },
      _react2.default.createElement(
        'div',
        { style: styles.compact, className: 'compact-picker ' + className },
        _react2.default.createElement(
          'div',
          null,
          (0, _map2.default)(colors, function (c) {
            return _react2.default.createElement(_CompactColor2.default, {
              key: c,
              color: c,
              active: c.toLowerCase() === hex,
              onClick: handleChange,
              onSwatchHover: onSwatchHover
            });
          }),
          _react2.default.createElement('div', { style: styles.clear })
        ),
        _react2.default.createElement(_CompactFields2.default, { hex: hex, rgb: rgb, onChange: handleChange })
      )
    );
  };
  
  Compact.propTypes = {
    colors: _propTypes2.default.arrayOf(_propTypes2.default.string)
  };
  
  Compact.defaultProps = {
    colors: ['#4D4D4D', '#999999', '#FFFFFF', '#F44E3B', '#FE9200', '#FCDC00', '#DBDF00', '#A4DD00', '#68CCCA', '#73D8FF', '#AEA1FF', '#FDA1FF', '#333333', '#808080', '#cccccc', '#D33115', '#E27300', '#FCC400', '#B0BC00', '#68BC00', '#16A5A5', '#009CE0', '#7B64FF', '#FA28FF', '#000000', '#666666', '#B3B3B3', '#9F0500', '#C45100', '#FB9E00', '#808900', '#194D33', '#0C797D', '#0062B1', '#653294', '#AB149E']
  };
  
  exports.default = (0, _common.ColorWrap)(Compact);

});
