// monaco-editor/esm/vs/editor/contrib/folding/foldingDecorations.js
define('node_modules/monaco-editor/esm/vs/editor/contrib/folding/foldingDecorations', function(require, exports, module) {

  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.FoldingDecorationProvider = void 0;
  /*---------------------------------------------------------------------------------------------
   *  Copyright (c) Microsoft Corporation. All rights reserved.
   *  Licensed under the MIT License. See License.txt in the project root for license information.
   *--------------------------------------------------------------------------------------------*/
  var textModel_js_1 = require("node_modules/monaco-editor/esm/vs/editor/common/model/textModel");
  var FoldingDecorationProvider = /** @class */ (function () {
      function FoldingDecorationProvider(editor) {
          this.editor = editor;
          this.autoHideFoldingControls = true;
      }
      FoldingDecorationProvider.prototype.getDecorationOption = function (isCollapsed) {
          if (isCollapsed) {
              return FoldingDecorationProvider.COLLAPSED_VISUAL_DECORATION;
          }
          else if (this.autoHideFoldingControls) {
              return FoldingDecorationProvider.EXPANDED_AUTO_HIDE_VISUAL_DECORATION;
          }
          else {
              return FoldingDecorationProvider.EXPANDED_VISUAL_DECORATION;
          }
      };
      FoldingDecorationProvider.prototype.deltaDecorations = function (oldDecorations, newDecorations) {
          return this.editor.deltaDecorations(oldDecorations, newDecorations);
      };
      FoldingDecorationProvider.prototype.changeDecorations = function (callback) {
          return this.editor.changeDecorations(callback);
      };
      FoldingDecorationProvider.COLLAPSED_VISUAL_DECORATION = textModel_js_1.ModelDecorationOptions.register({
          stickiness: 1 /* NeverGrowsWhenTypingAtEdges */,
          afterContentClassName: 'inline-folded',
          linesDecorationsClassName: 'folding collapsed'
      });
      FoldingDecorationProvider.EXPANDED_AUTO_HIDE_VISUAL_DECORATION = textModel_js_1.ModelDecorationOptions.register({
          stickiness: 1 /* NeverGrowsWhenTypingAtEdges */,
          linesDecorationsClassName: 'folding'
      });
      FoldingDecorationProvider.EXPANDED_VISUAL_DECORATION = textModel_js_1.ModelDecorationOptions.register({
          stickiness: 1 /* NeverGrowsWhenTypingAtEdges */,
          linesDecorationsClassName: 'folding alwaysShowFoldIcons'
      });
      return FoldingDecorationProvider;
  }());
  exports.FoldingDecorationProvider = FoldingDecorationProvider;
  

});
