// monaco-editor/esm/vs/base/browser/ui/findinput/findInputCheckboxes.js
define('node_modules/monaco-editor/esm/vs/base/browser/ui/findinput/findInputCheckboxes', function(require, exports, module) {

  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.RegexCheckbox = exports.WholeWordsCheckbox = exports.CaseSensitiveCheckbox = void 0;
  var tslib_1 = require("node_modules/tslib/tslib");
  /*---------------------------------------------------------------------------------------------
   *  Copyright (c) Microsoft Corporation. All rights reserved.
   *  Licensed under the MIT License. See License.txt in the project root for license information.
   *--------------------------------------------------------------------------------------------*/
  var __extends = (this && this.__extends) || (function () {
      var extendStatics = function (d, b) {
          extendStatics = Object.setPrototypeOf ||
              ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
              function (d, b) { for (var p in b)
                  if (b.hasOwnProperty(p))
                      d[p] = b[p]; };
          return extendStatics(d, b);
      };
      return function (d, b) {
          extendStatics(d, b);
          function __() { this.constructor = d; }
          d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
      };
  })();
  var checkbox_js_1 = require("node_modules/monaco-editor/esm/vs/base/browser/ui/checkbox/checkbox");
  ''/*@require node_modules/monaco-editor/esm/vs/base/browser/ui/findinput/findInputCheckboxes.css*/;
  var nls = tslib_1.__importStar(require("node_modules/monaco-editor/esm/vs/nls"));
  var NLS_CASE_SENSITIVE_CHECKBOX_LABEL = nls.localize('caseDescription', "Match Case");
  var NLS_WHOLE_WORD_CHECKBOX_LABEL = nls.localize('wordsDescription', "Match Whole Word");
  var NLS_REGEX_CHECKBOX_LABEL = nls.localize('regexDescription', "Use Regular Expression");
  var CaseSensitiveCheckbox = /** @class */ (function (_super) {
      __extends(CaseSensitiveCheckbox, _super);
      function CaseSensitiveCheckbox(opts) {
          return _super.call(this, {
              actionClassName: 'monaco-case-sensitive',
              title: NLS_CASE_SENSITIVE_CHECKBOX_LABEL + opts.appendTitle,
              isChecked: opts.isChecked,
              inputActiveOptionBorder: opts.inputActiveOptionBorder
          }) || this;
      }
      return CaseSensitiveCheckbox;
  }(checkbox_js_1.Checkbox));
  exports.CaseSensitiveCheckbox = CaseSensitiveCheckbox;
  var WholeWordsCheckbox = /** @class */ (function (_super) {
      __extends(WholeWordsCheckbox, _super);
      function WholeWordsCheckbox(opts) {
          return _super.call(this, {
              actionClassName: 'monaco-whole-word',
              title: NLS_WHOLE_WORD_CHECKBOX_LABEL + opts.appendTitle,
              isChecked: opts.isChecked,
              inputActiveOptionBorder: opts.inputActiveOptionBorder
          }) || this;
      }
      return WholeWordsCheckbox;
  }(checkbox_js_1.Checkbox));
  exports.WholeWordsCheckbox = WholeWordsCheckbox;
  var RegexCheckbox = /** @class */ (function (_super) {
      __extends(RegexCheckbox, _super);
      function RegexCheckbox(opts) {
          return _super.call(this, {
              actionClassName: 'monaco-regex',
              title: NLS_REGEX_CHECKBOX_LABEL + opts.appendTitle,
              isChecked: opts.isChecked,
              inputActiveOptionBorder: opts.inputActiveOptionBorder
          }) || this;
      }
      return RegexCheckbox;
  }(checkbox_js_1.Checkbox));
  exports.RegexCheckbox = RegexCheckbox;
  

});
