// amis/lib/renderers/Remark.js
define('node_modules/amis-editor/node_modules/amis/lib/renderers/Remark', function(require, exports, module) {

  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.RemarkRenderer = exports.filterContents = void 0;
  var tslib_1 = require("node_modules/tslib/tslib");
  var react_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/react/index"));
  var factory_1 = require("node_modules/amis-editor/node_modules/amis/lib/factory");
  var TooltipWrapper_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/TooltipWrapper"));
  var tpl_1 = require("node_modules/amis-editor/node_modules/amis/lib/utils/tpl");
  var theme_1 = require("node_modules/amis-editor/node_modules/amis/lib/theme");
  var icons_1 = require("node_modules/amis-editor/node_modules/amis/lib/components/icons");
  function filterContents(tooltip, data) {
      if (typeof tooltip === 'string') {
          return tpl_1.filter(tooltip, data);
      }
      else if (tooltip) {
          return tooltip.title
              ? {
                  title: tpl_1.filter(tooltip.title, data),
                  content: tooltip.content || tooltip.body
                      ? tpl_1.filter(tooltip.content || tooltip.body || '', data)
                      : undefined
              }
              : tooltip.content || tooltip.body
                  ? tpl_1.filter(tooltip.content || tooltip.body || '', data)
                  : undefined;
      }
      return tooltip;
  }
  exports.filterContents = filterContents;
  var Remark = /** @class */ (function (_super) {
      tslib_1.__extends(Remark, _super);
      function Remark() {
          return _super !== null && _super.apply(this, arguments) || this;
      }
      Remark.prototype.render = function () {
          var _a = this.props, className = _a.className, icon = _a.icon, tooltip = _a.tooltip, placement = _a.placement, rootClose = _a.rootClose, trigger = _a.trigger, container = _a.container, ns = _a.classPrefix, cx = _a.classnames, content = _a.content, data = _a.data;
          var finalIcon = (tooltip && tooltip.icon) || icon;
          return (react_1.default.createElement("div", { className: cx("Remark", (tooltip && tooltip.className) || className || "Remark--warning") },
              react_1.default.createElement(TooltipWrapper_1.default, { classPrefix: ns, classnames: cx, tooltip: filterContents(tooltip || content, data), tooltipClassName: tooltip && tooltip.tooltipClassName, placement: (tooltip && tooltip.placement) || placement, rootClose: (tooltip && tooltip.rootClose) || rootClose, trigger: (tooltip && tooltip.trigger) || trigger, container: container, delay: tooltip && tooltip.delay }, finalIcon ? (react_1.default.createElement("i", { className: cx('Remark-icon', finalIcon) })) : (react_1.default.createElement(icons_1.Icon, { icon: "question", className: "icon" })))));
      };
      Remark.propsList = [];
      Remark.defaultProps = {
          icon: '',
          trigger: ['hover', 'focus']
      };
      return Remark;
  }(react_1.default.Component));
  exports.default = theme_1.themeable(Remark);
  var RemarkRenderer = /** @class */ (function (_super) {
      tslib_1.__extends(RemarkRenderer, _super);
      function RemarkRenderer() {
          return _super !== null && _super.apply(this, arguments) || this;
      }
      RemarkRenderer = tslib_1.__decorate([
          factory_1.Renderer({
              test: /(^|\/)remark$/,
              name: 'remark'
          })
      ], RemarkRenderer);
      return RemarkRenderer;
  }(Remark));
  exports.RemarkRenderer = RemarkRenderer;
  

});
