// amis/lib/index.js
define('node_modules/amis/lib/index', function(require, exports, module) {

  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  var tslib_1 = require("node_modules/tslib/tslib");
  /** @license amis v1.0.15-beta.23
   *
   * Copyright Baidu
   *
   * This source code is licensed under the Apache license found in the
   * LICENSE file in the root directory of this source tree.
   */
  var factory_1 = require("node_modules/amis/lib/factory");
  exports.render = factory_1.render;
  exports.Renderer = factory_1.Renderer;
  exports.getRendererByName = factory_1.getRendererByName;
  exports.resolveRenderer = factory_1.resolveRenderer;
  exports.filterSchema = factory_1.filterSchema;
  exports.clearStoresCache = factory_1.clearStoresCache;
  exports.updateEnv = factory_1.updateEnv;
  var api_1 = require("node_modules/amis/lib/utils/api");
  exports.wrapFetcher = api_1.wrapFetcher;
  exports.buildApi = api_1.buildApi;
  var tpl_1 = require("node_modules/amis/lib/utils/tpl");
  exports.filter = tpl_1.filter;
  exports.reigsterTplEnginer = tpl_1.reigsterTplEnginer;
  exports.evalExpression = tpl_1.evalExpression;
  exports.evalJS = tpl_1.evalJS;
  exports.setCustomEvalJs = tpl_1.setCustomEvalJs;
  exports.setCustomEvalExpression = tpl_1.setCustomEvalExpression;
  var utils = tslib_1.__importStar(require("node_modules/amis/lib/utils/helper"));
  exports.utils = utils;
  var resize_sensor_1 = require("node_modules/amis/lib/utils/resize-sensor");
  exports.resizeSensor = resize_sensor_1.resizeSensor;
  var IconPickerIcons_1 = require("node_modules/amis/lib/renderers/Form/IconPickerIcons");
  exports.setIconVendor = IconPickerIcons_1.setIconVendor;
  var icons_1 = require("node_modules/amis/lib/components/icons");
  exports.Icon = icons_1.Icon;
  exports.registerIcon = icons_1.registerIcon;
  var store_1 = require("node_modules/amis/lib/store/index");
  exports.RegisterStore = store_1.RegisterStore;
  var locale_1 = require("node_modules/amis/lib/locale");
  exports.setDefaultLocale = locale_1.setDefaultLocale;
  exports.getDefaultLocale = locale_1.getDefaultLocale;
  exports.makeTranslator = locale_1.makeTranslator;
  exports.registerLocale = locale_1.register;
  // 注册渲染器
  require("node_modules/amis/lib/renderers/Action");
  require("node_modules/amis/lib/renderers/Alert");
  require("node_modules/amis/lib/renderers/Remark");
  require("node_modules/amis/lib/renderers/ButtonGroup");
  require("node_modules/amis/lib/renderers/ButtonToolbar");
  require("node_modules/amis/lib/renderers/DropDownButton");
  require("node_modules/amis/lib/renderers/Collapse");
  require("node_modules/amis/lib/renderers/Color");
  require("node_modules/amis/lib/renderers/CRUD");
  require("node_modules/amis/lib/renderers/Pagination");
  require("node_modules/amis/lib/renderers/Cards");
  require("node_modules/amis/lib/renderers/Card");
  require("node_modules/amis/lib/renderers/Date");
  require("node_modules/amis/lib/renderers/Dialog");
  require("node_modules/amis/lib/renderers/Divider");
  require("node_modules/amis/lib/renderers/Each");
  require("node_modules/amis/lib/renderers/Form/index");
  require("node_modules/amis/lib/renderers/Form/Control");
  require("node_modules/amis/lib/renderers/Form/Hidden");
  require("node_modules/amis/lib/renderers/Form/Text");
  require("node_modules/amis/lib/renderers/Form/Tag");
  require("node_modules/amis/lib/renderers/Form/Number");
  require("node_modules/amis/lib/renderers/Form/Textarea");
  require("node_modules/amis/lib/renderers/Form/Checkboxes");
  require("node_modules/amis/lib/renderers/Form/Checkbox");
  require("node_modules/amis/lib/renderers/Form/City");
  require("node_modules/amis/lib/renderers/Form/Rating");
  require("node_modules/amis/lib/renderers/Form/Switch");
  require("node_modules/amis/lib/renderers/Form/Button");
  require("node_modules/amis/lib/renderers/Form/ButtonGroup");
  require("node_modules/amis/lib/renderers/Form/ButtonToolbar");
  require("node_modules/amis/lib/renderers/Form/Radios");
  require("node_modules/amis/lib/renderers/Form/List");
  require("node_modules/amis/lib/renderers/Form/Location");
  require("node_modules/amis/lib/renderers/Form/Select");
  require("node_modules/amis/lib/renderers/Form/Static");
  require("node_modules/amis/lib/renderers/Form/Date");
  require("node_modules/amis/lib/renderers/Form/DateRange");
  require("node_modules/amis/lib/renderers/Form/Repeat");
  require("node_modules/amis/lib/renderers/Form/Tree");
  require("node_modules/amis/lib/renderers/Form/TreeSelect");
  require("node_modules/amis/lib/renderers/Form/Image");
  require("node_modules/amis/lib/renderers/Form/File");
  require("node_modules/amis/lib/renderers/Form/Matrix");
  require("node_modules/amis/lib/renderers/Form/Range");
  require("node_modules/amis/lib/renderers/Form/Array");
  require("node_modules/amis/lib/renderers/Form/Combo");
  require("node_modules/amis/lib/renderers/Form/Container");
  require("node_modules/amis/lib/renderers/Form/SubForm");
  require("node_modules/amis/lib/renderers/Form/RichText");
  require("node_modules/amis/lib/renderers/Form/Editor");
  require("node_modules/amis/lib/renderers/Form/DiffEditor");
  require("node_modules/amis/lib/renderers/Form/Grid");
  require("node_modules/amis/lib/renderers/Form/HBox");
  require("node_modules/amis/lib/renderers/Form/Panel");
  require("node_modules/amis/lib/renderers/Form/Color");
  require("node_modules/amis/lib/renderers/Form/ChainedSelect");
  require("node_modules/amis/lib/renderers/Form/NestedSelect");
  require("node_modules/amis/lib/renderers/Form/Transfer");
  require("node_modules/amis/lib/renderers/Form/Service");
  require("node_modules/amis/lib/renderers/Form/Table");
  require("node_modules/amis/lib/renderers/Form/Picker");
  require("node_modules/amis/lib/renderers/Form/IconPicker");
  require("node_modules/amis/lib/renderers/Form/Formula");
  require("node_modules/amis/lib/renderers/Form/FieldSet");
  require("node_modules/amis/lib/renderers/Form/Tabs");
  require("node_modules/amis/lib/renderers/Form/TabsTransfer");
  require("node_modules/amis/lib/renderers/Form/Group");
  require("node_modules/amis/lib/renderers/Form/InputGroup");
  require("node_modules/amis/lib/renderers/Grid");
  require("node_modules/amis/lib/renderers/HBox");
  require("node_modules/amis/lib/renderers/VBox");
  require("node_modules/amis/lib/renderers/Image");
  require("node_modules/amis/lib/renderers/Images");
  require("node_modules/amis/lib/renderers/List");
  require("node_modules/amis/lib/renderers/Operation");
  require("node_modules/amis/lib/renderers/Page");
  require("node_modules/amis/lib/renderers/Panel");
  require("node_modules/amis/lib/renderers/Plain");
  require("node_modules/amis/lib/renderers/Spinner");
  require("node_modules/amis/lib/renderers/Table");
  require("node_modules/amis/lib/renderers/Tabs");
  require("node_modules/amis/lib/renderers/Tpl");
  require("node_modules/amis/lib/renderers/Mapping");
  require("node_modules/amis/lib/renderers/Progress");
  require("node_modules/amis/lib/renderers/Status");
  require("node_modules/amis/lib/renderers/Json");
  require("node_modules/amis/lib/renderers/Link");
  require("node_modules/amis/lib/renderers/Switch");
  require("node_modules/amis/lib/renderers/Wizard");
  require("node_modules/amis/lib/renderers/Chart");
  require("node_modules/amis/lib/renderers/Container");
  require("node_modules/amis/lib/renderers/Service");
  require("node_modules/amis/lib/renderers/Video");
  require("node_modules/amis/lib/renderers/Audio");
  require("node_modules/amis/lib/renderers/Nav");
  require("node_modules/amis/lib/renderers/Tasks");
  require("node_modules/amis/lib/renderers/Drawer");
  require("node_modules/amis/lib/renderers/Wrapper");
  require("node_modules/amis/lib/renderers/IFrame");
  require("node_modules/amis/lib/renderers/QRCode");
  require("node_modules/amis/lib/renderers/Icon");
  require("node_modules/amis/lib/renderers/Carousel");
  var Scoped_1 = tslib_1.__importStar(require("node_modules/amis/lib/Scoped"));
  exports.Scoped = Scoped_1.default;
  exports.ScopedContext = Scoped_1.ScopedContext;
  var Item_1 = require("node_modules/amis/lib/renderers/Form/Item");
  exports.FormItem = Item_1.FormItem;
  // 兼容旧版本用法
  require("node_modules/amis/lib/compat");
  require("node_modules/amis/lib/themes/default");
  require("node_modules/amis/lib/themes/cxd");
  require("node_modules/amis/lib/themes/dark");
  var tpl_builtin_1 = require("node_modules/amis/lib/utils/tpl-builtin");
  exports.registerFilter = tpl_builtin_1.registerFilter;
  exports.filterDate = tpl_builtin_1.filterDate;
  exports.relativeValueRe = tpl_builtin_1.relativeValueRe;
  exports.resolveVariable = tpl_builtin_1.resolveVariable;
  exports.resolveVariableAndFilter = tpl_builtin_1.resolveVariableAndFilter;
  var validations_1 = require("node_modules/amis/lib/utils/validations");
  exports.addRule = validations_1.addRule;
  exports.str2rules = validations_1.str2rules;
  exports.validate = validations_1.validate;
  exports.validateObject = validations_1.validateObject;
  var Select_1 = require("node_modules/amis/lib/components/Select");
  exports.normalizeOptions = Select_1.normalizeOptions;
  var Options_1 = require("node_modules/amis/lib/renderers/Form/Options");
  exports.OptionsControl = Options_1.OptionsControl;
  var theme_1 = require("node_modules/amis/lib/theme");
  exports.classnames = theme_1.classnames;
  exports.getClassPrefix = theme_1.getClassPrefix;
  exports.setDefaultTheme = theme_1.setDefaultTheme;
  exports.theme = theme_1.theme;
  exports.getTheme = theme_1.getTheme;
  var classPrefix = theme_1.getClassPrefix();
  exports.classPrefix = classPrefix;
  tslib_1.__exportStar(require("node_modules/amis/lib/components/index"), exports);
  

});
