// amis/lib/components/index.js
define('node_modules/amis/lib/components/index', function(require, exports, module) {

  "use strict";
  /**
   * @file Index
   * @description
   * @author fex
   */
  Object.defineProperty(exports, "__esModule", { value: true });
  var tslib_1 = require("node_modules/tslib/tslib");
  var _404_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/404"));
  exports.NotFound = _404_1.default;
  var Alert_1 = tslib_1.__importStar(require("node_modules/amis/lib/components/Alert"));
  exports.AlertComponent = Alert_1.default;
  exports.alert = Alert_1.alert;
  exports.confirm = Alert_1.confirm;
  exports.prompt = Alert_1.prompt;
  var ContextMenu_1 = tslib_1.__importStar(require("node_modules/amis/lib/components/ContextMenu"));
  exports.ContextMenu = ContextMenu_1.default;
  exports.openContextMenus = ContextMenu_1.openContextMenus;
  var AsideNav_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/AsideNav"));
  exports.AsideNav = AsideNav_1.default;
  var Button_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Button"));
  exports.Button = Button_1.default;
  var Checkbox_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Checkbox"));
  exports.Checkbox = Checkbox_1.default;
  var Checkboxes_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Checkboxes"));
  exports.Checkboxes = Checkboxes_1.default;
  var Collapse_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Collapse"));
  exports.Collapse = Collapse_1.default;
  var ColorPicker_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/ColorPicker"));
  exports.ColorPicker = ColorPicker_1.default;
  var DatePicker_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/DatePicker"));
  exports.DatePicker = DatePicker_1.default;
  var DateRangePicker_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/DateRangePicker"));
  exports.DateRangePicker = DateRangePicker_1.default;
  var Drawer_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Drawer"));
  exports.Drawer = Drawer_1.default;
  var Tabs_1 = tslib_1.__importStar(require("node_modules/amis/lib/components/Tabs"));
  exports.Tabs = Tabs_1.default;
  exports.Tab = Tabs_1.Tab;
  var Editor_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Editor"));
  exports.Editor = Editor_1.default;
  var Html_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Html"));
  exports.Html = Html_1.default;
  var Icons = tslib_1.__importStar(require("node_modules/amis/lib/components/icons"));
  exports.Icons = Icons;
  var Layout_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Layout"));
  exports.Layout = Layout_1.default;
  var LazyComponent_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/LazyComponent"));
  exports.LazyComponent = LazyComponent_1.default;
  var Modal_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Modal"));
  exports.Modal = Modal_1.default;
  var Overlay_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Overlay"));
  exports.Overlay = Overlay_1.default;
  var PopOver_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/PopOver"));
  exports.PopOver = PopOver_1.default;
  var Radios_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Radios"));
  exports.Radios = Radios_1.default;
  var Range_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Range"));
  exports.Range = Range_1.default;
  var Rating_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Rating"));
  exports.Rating = Rating_1.default;
  // import RichText from './RichText';
  var Select_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Select"));
  exports.Select = Select_1.default;
  var Spinner_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Spinner"));
  exports.Spinner = Spinner_1.default;
  var Switch_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Switch"));
  exports.Switch = Switch_1.default;
  var Textarea_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Textarea"));
  exports.Textarea = Textarea_1.default;
  var TitleBar_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/TitleBar"));
  exports.TitleBar = TitleBar_1.default;
  var Toast_1 = tslib_1.__importStar(require("node_modules/amis/lib/components/Toast"));
  exports.ToastComponent = Toast_1.default;
  exports.toast = Toast_1.toast;
  var Tooltip_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Tooltip"));
  exports.Tooltip = Tooltip_1.default;
  var TooltipWrapper_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/TooltipWrapper"));
  exports.TooltipWrapper = TooltipWrapper_1.default;
  var Tree_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Tree"));
  exports.Tree = Tree_1.default;
  var Alert2_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Alert2"));
  exports.Alert2 = Alert2_1.default;
  var Transfer_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/Transfer"));
  exports.Transfer = Transfer_1.default;
  var TabsTransfer_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/TabsTransfer"));
  exports.TabsTransfer = TabsTransfer_1.default;
  var ListCheckboxes_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/ListCheckboxes"));
  exports.ListCheckboxes = ListCheckboxes_1.default;
  var TableCheckboxes_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/TableCheckboxes"));
  exports.TableCheckboxes = TableCheckboxes_1.default;
  var TreeCheckboxes_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/TreeCheckboxes"));
  exports.TreeCheckboxes = TreeCheckboxes_1.default;
  var ChainedCheckboxes_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/ChainedCheckboxes"));
  exports.ChainedCheckboxes = ChainedCheckboxes_1.default;
  var ResultBox_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/ResultBox"));
  exports.ResultBox = ResultBox_1.default;
  var InputBox_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/InputBox"));
  exports.InputBox = InputBox_1.default;
  var ListRadios_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/ListRadios"));
  exports.ListRadios = ListRadios_1.default;
  var TreeRadios_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/TreeRadios"));
  exports.TreeRadios = TreeRadios_1.default;
  var ListGroup_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/ListGroup"));
  exports.ListGroup = ListGroup_1.default;
  var NumberInput_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/NumberInput"));
  exports.NumberInput = NumberInput_1.default;
  var ArrayInput_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/ArrayInput"));
  exports.ArrayInput = ArrayInput_1.default;
  var SearchBox_1 = tslib_1.__importDefault(require("node_modules/amis/lib/components/SearchBox"));
  exports.SearchBox = SearchBox_1.default;
  

});
