// amis/lib/components/index.js
define('node_modules/amis-editor/node_modules/amis/lib/components/index', function(require, exports, module) {

  "use strict";
  /**
   * @file Index
   * @description
   * @author fex
   */
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.ArrayInput = exports.NumberInput = exports.ListGroup = exports.TreeRadios = exports.SearchBox = exports.ListRadios = exports.InputBox = exports.ResultBox = exports.ChainedCheckboxes = exports.TreeCheckboxes = exports.TableCheckboxes = exports.ListCheckboxes = exports.TabsTransfer = exports.Transfer = exports.Tree = exports.TooltipWrapper = exports.Tooltip = exports.toast = exports.ToastComponent = exports.TitleBar = exports.Textarea = exports.Switch = exports.Spinner = exports.Select = exports.Rating = exports.Range = exports.Radios = exports.PopOver = exports.Overlay = exports.Modal = exports.LazyComponent = exports.Layout = exports.Icons = exports.Html = exports.Editor = exports.Tab = exports.Tabs = exports.Drawer = exports.DateRangePicker = exports.DatePicker = exports.ColorPicker = exports.Collapse = exports.Checkboxes = exports.Checkbox = exports.Button = exports.AsideNav = exports.Alert2 = exports.openContextMenus = exports.ContextMenu = exports.prompt = exports.confirm = exports.alert = exports.AlertComponent = exports.NotFound = void 0;
  var tslib_1 = require("node_modules/tslib/tslib");
  var _404_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/404"));
  exports.NotFound = _404_1.default;
  var Alert_1 = tslib_1.__importStar(require("node_modules/amis-editor/node_modules/amis/lib/components/Alert"));
  Object.defineProperty(exports, "AlertComponent", { enumerable: true, get: function () { return Alert_1.default; } });
  Object.defineProperty(exports, "alert", { enumerable: true, get: function () { return Alert_1.alert; } });
  Object.defineProperty(exports, "confirm", { enumerable: true, get: function () { return Alert_1.confirm; } });
  Object.defineProperty(exports, "prompt", { enumerable: true, get: function () { return Alert_1.prompt; } });
  var ContextMenu_1 = tslib_1.__importStar(require("node_modules/amis-editor/node_modules/amis/lib/components/ContextMenu"));
  Object.defineProperty(exports, "ContextMenu", { enumerable: true, get: function () { return ContextMenu_1.default; } });
  Object.defineProperty(exports, "openContextMenus", { enumerable: true, get: function () { return ContextMenu_1.openContextMenus; } });
  var AsideNav_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/AsideNav"));
  exports.AsideNav = AsideNav_1.default;
  var Button_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Button"));
  exports.Button = Button_1.default;
  var Checkbox_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Checkbox"));
  exports.Checkbox = Checkbox_1.default;
  var Checkboxes_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Checkboxes"));
  exports.Checkboxes = Checkboxes_1.default;
  var Collapse_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Collapse"));
  exports.Collapse = Collapse_1.default;
  var ColorPicker_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/ColorPicker"));
  exports.ColorPicker = ColorPicker_1.default;
  var DatePicker_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/DatePicker"));
  exports.DatePicker = DatePicker_1.default;
  var DateRangePicker_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/DateRangePicker"));
  exports.DateRangePicker = DateRangePicker_1.default;
  var Drawer_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Drawer"));
  exports.Drawer = Drawer_1.default;
  var Tabs_1 = tslib_1.__importStar(require("node_modules/amis-editor/node_modules/amis/lib/components/Tabs"));
  Object.defineProperty(exports, "Tabs", { enumerable: true, get: function () { return Tabs_1.default; } });
  Object.defineProperty(exports, "Tab", { enumerable: true, get: function () { return Tabs_1.Tab; } });
  var Editor_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Editor"));
  exports.Editor = Editor_1.default;
  var Html_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Html"));
  exports.Html = Html_1.default;
  var Icons = tslib_1.__importStar(require("node_modules/amis-editor/node_modules/amis/lib/components/icons"));
  exports.Icons = Icons;
  var Layout_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Layout"));
  exports.Layout = Layout_1.default;
  var LazyComponent_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/LazyComponent"));
  exports.LazyComponent = LazyComponent_1.default;
  var Modal_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Modal"));
  exports.Modal = Modal_1.default;
  var Overlay_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Overlay"));
  exports.Overlay = Overlay_1.default;
  var PopOver_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/PopOver"));
  exports.PopOver = PopOver_1.default;
  var Radios_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Radios"));
  exports.Radios = Radios_1.default;
  var Range_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Range"));
  exports.Range = Range_1.default;
  var Rating_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Rating"));
  exports.Rating = Rating_1.default;
  // import RichText from './RichText';
  var Select_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Select"));
  exports.Select = Select_1.default;
  var Spinner_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Spinner"));
  exports.Spinner = Spinner_1.default;
  var Switch_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Switch"));
  exports.Switch = Switch_1.default;
  var Textarea_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Textarea"));
  exports.Textarea = Textarea_1.default;
  var TitleBar_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/TitleBar"));
  exports.TitleBar = TitleBar_1.default;
  var Toast_1 = tslib_1.__importStar(require("node_modules/amis-editor/node_modules/amis/lib/components/Toast"));
  Object.defineProperty(exports, "ToastComponent", { enumerable: true, get: function () { return Toast_1.default; } });
  Object.defineProperty(exports, "toast", { enumerable: true, get: function () { return Toast_1.toast; } });
  var Tooltip_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Tooltip"));
  exports.Tooltip = Tooltip_1.default;
  var TooltipWrapper_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/TooltipWrapper"));
  exports.TooltipWrapper = TooltipWrapper_1.default;
  var Tree_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Tree"));
  exports.Tree = Tree_1.default;
  var Alert2_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Alert2"));
  exports.Alert2 = Alert2_1.default;
  var Transfer_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/Transfer"));
  exports.Transfer = Transfer_1.default;
  var TabsTransfer_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/TabsTransfer"));
  exports.TabsTransfer = TabsTransfer_1.default;
  var ListCheckboxes_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/ListCheckboxes"));
  exports.ListCheckboxes = ListCheckboxes_1.default;
  var TableCheckboxes_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/TableCheckboxes"));
  exports.TableCheckboxes = TableCheckboxes_1.default;
  var TreeCheckboxes_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/TreeCheckboxes"));
  exports.TreeCheckboxes = TreeCheckboxes_1.default;
  var ChainedCheckboxes_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/ChainedCheckboxes"));
  exports.ChainedCheckboxes = ChainedCheckboxes_1.default;
  var ResultBox_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/ResultBox"));
  exports.ResultBox = ResultBox_1.default;
  var InputBox_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/InputBox"));
  exports.InputBox = InputBox_1.default;
  var ListRadios_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/ListRadios"));
  exports.ListRadios = ListRadios_1.default;
  var TreeRadios_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/TreeRadios"));
  exports.TreeRadios = TreeRadios_1.default;
  var ListGroup_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/ListGroup"));
  exports.ListGroup = ListGroup_1.default;
  var NumberInput_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/NumberInput"));
  exports.NumberInput = NumberInput_1.default;
  var ArrayInput_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/ArrayInput"));
  exports.ArrayInput = ArrayInput_1.default;
  var SearchBox_1 = tslib_1.__importDefault(require("node_modules/amis-editor/node_modules/amis/lib/components/SearchBox"));
  exports.SearchBox = SearchBox_1.default;
  

});
