// react-datetime/src/CalendarContainer.js
define('node_modules/amis-editor/node_modules/react-datetime/src/CalendarContainer', function(require, exports, module) {

  'use strict';
  
  var React = require('node_modules/amis-editor/node_modules/react/index'),
  	createClass = require('node_modules/create-react-class/index'),
  	DaysView = require('node_modules/amis-editor/node_modules/react-datetime/src/DaysView'),
  	MonthsView = require('node_modules/amis-editor/node_modules/react-datetime/src/MonthsView'),
  	YearsView = require('node_modules/amis-editor/node_modules/react-datetime/src/YearsView'),
  	TimeView = require('node_modules/amis-editor/node_modules/react-datetime/src/TimeView')
  	;
  
  var CalendarContainer = createClass({
  	viewComponents: {
  		days: DaysView,
  		months: MonthsView,
  		years: YearsView,
  		time: TimeView
  	},
  
  	render: function() {
  		return React.createElement( this.viewComponents[ this.props.view ], this.props.viewProps );
  	}
  });
  
  module.exports = CalendarContainer;
  

});
