// react-json-tree/lib/JSONArrayNode.js
define('node_modules/react-json-tree/lib/JSONArrayNode', function(require, exports, module) {

  'use strict';
  
  exports.__esModule = true;
  
  var _extends2 = require('node_modules/babel-runtime/helpers/extends');
  
  var _extends3 = _interopRequireDefault(_extends2);
  
  var _objectWithoutProperties2 = require('node_modules/babel-runtime/helpers/objectWithoutProperties');
  
  var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);
  
  var _react = require('node_modules/react/index');
  
  var _react2 = _interopRequireDefault(_react);
  
  var _propTypes = require('node_modules/prop-types/index');
  
  var _propTypes2 = _interopRequireDefault(_propTypes);
  
  var _JSONNestedNode = require('node_modules/react-json-tree/lib/JSONNestedNode');
  
  var _JSONNestedNode2 = _interopRequireDefault(_JSONNestedNode);
  
  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }
  
  // Returns the "n Items" string for this node,
  // generating and caching it if it hasn't been created yet.
  function createItemString(data) {
    return data.length + ' ' + (data.length !== 1 ? 'items' : 'item');
  }
  
  // Configures <JSONNestedNode> to render an Array
  var JSONArrayNode = function JSONArrayNode(_ref) {
    var data = _ref.data,
        props = (0, _objectWithoutProperties3['default'])(_ref, ['data']);
    return _react2['default'].createElement(_JSONNestedNode2['default'], (0, _extends3['default'])({}, props, {
      data: data,
      nodeType: 'Array',
      nodeTypeIndicator: '[]',
      createItemString: createItemString,
      expandable: data.length > 0
    }));
  };
  
  JSONArrayNode.propTypes = {
    data: _propTypes2['default'].array
  };
  
  exports['default'] = JSONArrayNode;

});
