// react-json-tree/lib/JSONObjectNode.js
define('node_modules/react-json-tree/lib/JSONObjectNode', function(require, exports, module) {

  'use strict';
  
  exports.__esModule = true;
  
  var _extends2 = require('node_modules/babel-runtime/helpers/extends');
  
  var _extends3 = _interopRequireDefault(_extends2);
  
  var _objectWithoutProperties2 = require('node_modules/babel-runtime/helpers/objectWithoutProperties');
  
  var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);
  
  var _getOwnPropertyNames = require('node_modules/babel-runtime/core-js/object/get-own-property-names');
  
  var _getOwnPropertyNames2 = _interopRequireDefault(_getOwnPropertyNames);
  
  var _react = require('node_modules/react/index');
  
  var _react2 = _interopRequireDefault(_react);
  
  var _propTypes = require('node_modules/prop-types/index');
  
  var _propTypes2 = _interopRequireDefault(_propTypes);
  
  var _JSONNestedNode = require('node_modules/react-json-tree/lib/JSONNestedNode');
  
  var _JSONNestedNode2 = _interopRequireDefault(_JSONNestedNode);
  
  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }
  
  // Returns the "n Items" string for this node,
  // generating and caching it if it hasn't been created yet.
  function createItemString(data) {
    var len = (0, _getOwnPropertyNames2['default'])(data).length;
    return len + ' ' + (len !== 1 ? 'keys' : 'key');
  }
  
  // Configures <JSONNestedNode> to render an Object
  var JSONObjectNode = function JSONObjectNode(_ref) {
    var data = _ref.data,
        props = (0, _objectWithoutProperties3['default'])(_ref, ['data']);
    return _react2['default'].createElement(_JSONNestedNode2['default'], (0, _extends3['default'])({}, props, {
      data: data,
      nodeType: 'Object',
      nodeTypeIndicator: props.nodeType === 'Error' ? 'Error()' : '{}',
      createItemString: createItemString,
      expandable: (0, _getOwnPropertyNames2['default'])(data).length > 0
    }));
  };
  
  JSONObjectNode.propTypes = {
    data: _propTypes2['default'].object,
    nodeType: _propTypes2['default'].string
  };
  
  exports['default'] = JSONObjectNode;

});
