// monaco-editor/esm/vs/base/browser/dnd.js
define('node_modules/monaco-editor/esm/vs/base/browser/dnd', function(require, exports, module) {

  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.StaticDND = exports.DragAndDropData = exports.DataTransfers = void 0;
  // Common data transfers
  exports.DataTransfers = {
      /**
       * Application specific resource transfer type
       */
      RESOURCES: 'ResourceURLs',
      /**
       * Browser specific transfer type to download
       */
      DOWNLOAD_URL: 'DownloadURL',
      /**
       * Browser specific transfer type for files
       */
      FILES: 'Files',
      /**
       * Typically transfer type for copy/paste transfers.
       */
      TEXT: 'text/plain'
  };
  var DragAndDropData = /** @class */ (function () {
      function DragAndDropData(data) {
          this.data = data;
      }
      DragAndDropData.prototype.update = function () {
          // noop
      };
      DragAndDropData.prototype.getData = function () {
          return this.data;
      };
      return DragAndDropData;
  }());
  exports.DragAndDropData = DragAndDropData;
  exports.StaticDND = {
      CurrentDragAndDropData: undefined
  };
  

});
